Reenable compiler warning 4244 under coreclr/

AaronRobinsonMSFT opened the pull request:
Contributes to https://github.com/dotnet/runtime/issues/66154

Reenable and fix MSVC compiler warnings under `coreclr/`.

/cc @GrabYourPitchforks @jkotas @janvorli

AaronRobinsonMSFT committed:
From vm/

AaronRobinsonMSFT committed:
From util/

AaronRobinsonMSFT committed:
From debug/

AaronRobinsonMSFT committed:
GCInfo related

jkotas commented on src/coreclr/vm/typehash.cpp:
```diff
@@ -234,10 +229,13 @@ static DWORD HashTypeHandle(TypeHandle t)
         _ASSERTE(!"Generic variables are unexpected here.");
-        retVal = t.AsTAddr();
+        // Revisit the cast to DWORD if this path becomes viable.
+        retVal = (DWORD)t.AsTAddr();
```
It would be more appropriate to return `0` as fallback hashcode here.

AaronRobinsonMSFT replied:
Good idea.

AaronRobinsonMSFT commented on src/coreclr/jit/jitgcinfo.h:
```diff
@@ -20,14 +20,14 @@
 struct RegSlotIdKey
 {
-    unsigned short m_regNum;
+    unsigned m_regNum;
```
I wanted to, but am unclear how to confirm the number of regs in the target scenario can fit inside a short. The `regNumber` type is defined to be:

```c++
enum _regNumber_enum : unsigned
{
#define REGDEF(name, rnum, mask, sname) REG_##name = rnum,
#define REGALIAS(alias, realname) REG_##alias = REG_##realname,
#include "register.h"

    REG_COUNT,
    REG_NA           = REG_COUNT,
    ACTUAL_REG_COUNT = REG_COUNT - 1 // everything but REG_STK (only real regs)
};
```

Usage is below. I _think_ the register usage is safe based on words like "small", but unclear how to assert to confirm that.

https://github.com/dotnet/runtime/blob/c3ef9da5a33b64d35269e3d780525a57ccc56ee0/src/coreclr/jit/gcencode.cpp#L4653-L4667.

jkotas commented on src/coreclr/inc/gcinfotypes.h:
```diff
@@ -53,7 +53,7 @@ inline UINT32 CeilOfLog2(size_t x)
 }
 
-enum GcSlotFlags
+enum GcSlotFlags : unsigned short
```
Is this fixing any warnings, or is it just an unrelated change?

AaronRobinsonMSFT replied:
This is fixing warnings on usage. One of the hashing mechanisms is assumes the flags will fit into an `unsigned short` and is types as such. By defining it here that assumption flows nicely.

jkotas replied:
Yes, but it is creating inconsistencies in the enum definitions. We tend to prefer to use the default enum underlying type. It makes this enum to look different from the other enums next to it. And there is mirror of this enum in C# that would need to be updated too. 

If it is just one place in hashing, I would add a cast there instead.

AaronRobinsonMSFT replied:
I would much rather properly define the type to be honest. The expressiveness exists so I'd argue we should use it. I see how many things would need to change.

AaronRobinsonMSFT replied:
There would be 7 seemingly random casts in the `gcencode.cpp` file. I really dislike this design here. We are adding far more confusion than clarity when we should be expressing our intent clearly with the type-system that exists.

AaronRobinsonMSFT replied:
I think I have a compromise.

jkotas replied:
FWIW, https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Renum-underlying says "Specify the underlying type of an enumeration only when necessary". It is not obvious to me that avoiding one cast in hashing qualifies as necessary.

AaronRobinsonMSFT replied:
Agreed for `enum class`. I don't see the same argument for `enum` types that grow and transparently cast to the underlying type. The `enum class` doesn't permit this so casting is required and thus frowned upon.

AaronRobinsonMSFT committed:
Apply review feedback.

Use MAXDWORD instead of DWORD_MAX.
Remove DWORD_MAX from pal, since it is not used anywhere.

AaronRobinsonMSFT committed:
Apply JIT format

AaronRobinsonMSFT commented:
@jkotas Any other concerns?

jkotas reviewed (approved):
LGTM. Thank you for working on this!
